Modules/_abc.c

```diff
@@ -0,0 +1,960 @@
+_add_to_registry(PyObject *self, PyObject *cls)
+{
+    _abc_data *impl;
+    int res;
+    impl = _get_impl(self);
+    if (!impl) {
+        return -1;
+    }
+    res = _add_to_weak_set((PyObject *)(impl->_abc_registry), cls, 1);
+    Py_DECREF(impl);
+    return res;
+}
+
+static int
+_add_to_cache(_abc_data *impl, PyObject *cls)
+{
+    return _add_to_weak_set(impl->_abc_cache, cls, 0);
+}
+
+static int
+_add_to_negative_cache(PyObject *self, PyObject *cls)
+{
+    _abc_data *impl;
+    int res;
+    impl = _get_impl(self);
+    if (!impl) {
+        return -1;
+    }
+    res = _add_to_weak_set(impl->_abc_negative_cache, cls, 0);
+    Py_DECREF(impl);
+    return res;
+}
+
+PyDoc_STRVAR(_reset_registry_doc,
+"Internal ABC helper to reset registry of a given class.\n\
+\n\
+Should be only used by refleak.py");
+
+static PyObject *
+_reset_registry(PyObject *m, PyObject *args)
+{
+    PyObject *self, *registry;
+    if (!PyArg_UnpackTuple(args, "_reset_registry", 1, 1, &self)) {
+        return NULL;
+    }
+    registry = _get_registry(self);
+    if (!registry) {
+        return NULL;
+    }
+    if (PySet_Clear(registry) < 0) {
+        return NULL;
+    }
+    Py_RETURN_NONE;
+}
+
+static int
+_reset_negative_cache(_abc_data *impl)
+{
+    return PySet_Clear(impl->_abc_negative_cache);
+}
+
+PyDoc_STRVAR(_reset_caches_doc,
+"Internal ABC helper to reset both caches of a given class.\n\
+\n\
+Should be only used by refleak.py");
+
+static PyObject *
+_reset_caches(PyObject *m, PyObject *args)
+{
+    PyObject *self;
+    _abc_data *impl;
+    if (!PyArg_UnpackTuple(args, "_reset_caches", 1, 1, &self)) {
+        return NULL;
+    }
+    impl = _get_impl(self);
+    if (!impl) {
+        return NULL;
+    }
+    if (PySet_Clear(impl->_abc_cache) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    /* also the second cache */
+    if (_reset_negative_cache(impl) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    Py_DECREF(impl);
+    Py_RETURN_NONE;
+}
+
+static PyObject *
+_get_negative_cache_version(_abc_data *impl)
+{
+    return impl->_abc_negative_cache_version;
+}
+
+static void
+_set_negative_cache_version(_abc_data *impl)
+{
+    Py_INCREF(abc_invalidation_counter);
+    Py_SETREF(impl->_abc_negative_cache_version, abc_invalidation_counter);
+}
+
+PyDoc_STRVAR(_get_dump_doc,
+"Internal ABC helper for cache and registry debugging.\n\
+\n\
+Return shallow copies of registry, of both caches, and\n\
+negative cache version. Don't call this function directly,\n\
+instead use ABC._dump_registry() for a nice repr.");
+
+static PyObject *
+_get_dump(PyObject *m, PyObject *args)
+{
+    PyObject *self, *registry, *cache, *negative_cache, *cache_version;
+    _abc_data *impl;
+    if (!PyArg_UnpackTuple(args, "_get_dump", 1, 1, &self)) {
+        return NULL;
+    }
+    registry = _get_registry(self);
+    if (!registry) {
+        return NULL;
+    }
+    registry = PyObject_CallMethod(registry, "copy", NULL);
```

Comment:
Is this a borrowed ref?  Please add `/* borrowed */` if so, otherwise it's hard to read the code.  I assumed that a few DECREFs are missing below.